### ISSUE TITLE
Define reference-oracle precedence for degenerate Kimi-K3 top-k routing

### ISSUE THREAD

**Issue by voipmonitor** [lukealonso/b12x#225]
## Status

**Research-only policy decision.** The Kimi-K3 batched top-16 implementation in
[B12X PR #224](https://github.com/local-inference-lab/b12x/pull/224) preserves
unique expert selection. This issue requests agreement on how review automation
must treat a reference implementation that violates the operation contract on
a degenerate input.

## Affected implementation

The affected production kernel and its qualification tooling are:

- `b12x/comm/pcie/_dcp_a2a_cute.py`: CuTeDSL batched Kimi-K3 top-16 selection.
- `b12x/comm/pcie/pcie_dcp_a2a.py`: caller-facing dispatch and input validation.
- `benchmarks/benchmark_kimi_k3_pair_topk.py`: vLLM parity and adversarial
  contract checks.
- `docs/evidence/20260816-kimi-k3-topk-tp16-batch8.json`: GPU correctness,
  graph-replay, provenance, and timing evidence.
- B12X tested revision: `038117b`.
- vLLM reference revision: `ad848fc4141f201489db18d5453c50b312245a0a`.

## Reference defect

The top-k operation selects 16 distinct expert positions without replacement.
A selected expert must not become eligible during a later selection round.

The following reachable reference input violates that invariant:

- rows: 8
- experts per row: 896
- selected experts per row: 16
- correction bias: `-inf` for every expert
- reference function: vLLM `vllm_topk_sigmoid`

vLLM returns the following result for every row:

- expert IDs:
  `[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]`
- weights: sixteen values equal to `0.0625`

The expert IDs are deterministic but invalid because one expert occupies all
16 top-k slots.

B12X explicitly tracks active experts and returns 16 distinct in-range IDs. It
matches vLLM exactly for finite random inputs, finite exact ties, and near-ties.

## Review-policy conflict

CodeRabbit produced two incompatible requirements on PR #224:

1. It correctly identified that score-only invalidation could reselect an
   expert when all candidate scores are `-inf`. The accepted correction added
   an explicit active-expert mask and enforced 16 distinct IDs.
2. It then required `ids_exact` for the `all_neg_inf` case because selected
   IDs are part of the vLLM parity contract.

Implementing the second requirement would restore the duplicate-expert defect
rejected by the first requirement. Exact equality is not a valid correctness
oracle when the reference output violates the declared operation invariant.

The repository review instructions require correctness before performance and
exact oracle validation, but they do not define which requirement takes
precedence when the oracle itself is invalid. CodeRabbit therefore interprets
reference equality as unconditional.

## Proposed oracle hierarchy

Use the following precedence for automated and human review:

1. **Declared semantic contract:** shape, range, uniqueness, normalization,
   dtype, synchronization, and serving-lifecycle invariants.
2. **Exact reference parity:** required for every input where the reference
   satisfies the declared contract.
3. **Documented reference divergence:** permitted only with a minimal
   reproducer proving that the reference violates a fundamental invariant.
4. **Deterministic degenerate behavior:** the implementation must define and
   test deterministic behavior for a reachable degenerate input even when exact
   reference parity is invalid.
5. **Performance evidence:** interpreted only after the applicable correctness
   checks pass.

Proposed `.coderabbit.yaml` instruction:

```text
Treat exact parity with a reference implementation as required only when the
reference output satisfies the declared operation contract. If the reference
violates a fundamental invariant, require a minimal reproducer, preserve the
invariant, document the intentional divergence, and require exact parity for
the reference's valid input domain. Do not request reproduction of a reference
defect solely to obtain exact equality.
```

### Reference-exact cases

Require exact expert IDs and weights for:

- finite random scores;
- finite exact ties;
- finite near-ties;
- mixed non-finite values when the reference still returns 16 distinct valid
  IDs.

### Semantic-adversarial cases

For `all_neg_inf` and any equivalent NaN-to-`-inf` mapping, require:

- 16 distinct expert IDs;
- every ID in `[0, 896)`;
- deterministic tie ordering;
- finite, positive FP32 weights;
- weight sum equal to one within the frozen tolerance;
- successful CUDA Graph capture and replay;
- stable caller-owned output addresses;
- no replay allocation.

The qualification receipt should record that vLLM exact-ID parity is
inapplicable because the reference violates uniqueness. It must not report
exact parity for this case.

## Decision requested

Please confirm whether B12X should:

- preserve unique deterministic top-k behavior and adopt the proposed oracle
  hierarchy; or
- reproduce vLLM's duplicate expert IDs for `all_neg_inf`.

The first option is recommended because it preserves the operation contract,
retains exact vLLM parity across the valid reference domain, and avoids
reintroducing a verified defect.

**Comment by voipmonitor** [comment]
## CuTeDSL implementation status

The mergeable Kimi-K3 router implementation contains no native CUDA, C++, or binary extension:

1. [B12X #224](https://github.com/local-inference-lab/b12x/pull/224) implements communication-backed and stateless top-16 selection in CuTeDSL.
2. [vLLM #386](https://github.com/local-inference-lab/vllm/pull/386) integrates projection transport and consumes a routing payload when the transport pool supports the token batch.
3. [vLLM #400](https://github.com/local-inference-lab/vllm/pull/400) invokes stateless CuTeDSL selection when vLLM already owns complete router logits.

B12X #198, which contained `pcie_vocab_argmax.cu`, and vLLM #317, which depended on it, are closed and are not part of the merge chain.

The native top-16 proof of concept measured about 6.97 microseconds. The CuTeDSL kernel measured 6.240 microseconds at median in a DSpark target trace and 6.975 microseconds at median in a DFlash target trace. The available evidence therefore does not justify retaining native CUDA.

The complete-tie oracle decision described by this issue is independent of the implementation language and does not block review of finite, tied-finite, near-tied, mixed-nonfinite, CUDA Graph, or runtime performance evidence.
